pulsar-client-cpp/lib/ReaderImpl.cc

```diff
@@ -0,0 +1,105 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "ClientImpl.h"
+#include "ReaderImpl.h"
+
+namespace pulsar {
+
+static ResultCallback emptyCallback;
+
+ReaderImpl::ReaderImpl(const ClientImplPtr client, const std::string& topic,
+                       const ReaderConfiguration& conf, const ExecutorServicePtr listenerExecutor,
+                       ReaderCallback readerCreatedCallback) :
+                topic_(topic),
+                    client_(client),
+                    readerConf_(conf),
+                    readerCreatedCallback_(readerCreatedCallback) {
+}
+
+void ReaderImpl::start(const BatchMessageId& startMessageId) {
+    ConsumerConfiguration consumerConf;
+    consumerConf.setConsumerType(ConsumerExclusive);
+    consumerConf.setReceiverQueueSize(readerConf_.getReceiverQueueSize());
+
+    if (readerConf_.getReaderName().length() > 0) {
+        consumerConf.setConsumerName(readerConf_.getReaderName());
+    }
+
+    if (readerConf_.hasReaderListener()) {
+        // Adapt the message listener to be a reader-listener
+        consumerConf.setMessageListener(
+                boost::bind(readerConf_.getReaderListener(), Reader(shared_from_this()), _2));
+    }
+
+    std::string subscription = "reader-" + generateRandomName();
+
+    consumer_ = boost::make_shared<ConsumerImpl>(client_.lock(), topic_, subscription, consumerConf,
+                                                 ExecutorServicePtr(), NonPartitioned,
+                                                 Commands::SubscriptionModeNonDurable,
+                                                 Optional<BatchMessageId>::of(startMessageId));
+    consumer_->getConsumerCreatedFuture().addListener(
+            boost::bind(&ReaderImpl::handleConsumerCreated, shared_from_this(), _1, _2));
+    consumer_->start();
+}
+
+const std::string& ReaderImpl::getTopic() const {
+    return consumer_->getTopic();
+}
+
+void ReaderImpl::handleConsumerCreated(Result result, ConsumerImplBaseWeakPtr consumer) {
+    readerCreatedCallback_(result, Reader(shared_from_this()));
+}
+
+ConsumerImplPtr ReaderImpl::getConsumer() {
+    return consumer_;
+}
+
+Result ReaderImpl::readNext(Message& msg) {
+    Result res = consumer_->receive(msg);
+    acknowledgeIfNecessary(res, msg);
+    return res;
+}
+
+Result ReaderImpl::readNext(Message& msg, int timeoutMs) {
+    Result res = consumer_->receive(msg, timeoutMs);
+    acknowledgeIfNecessary(res, msg);
+    return res;
+}
+
+void ReaderImpl::acknowledgeIfNecessary(Result result, const Message& msg) {
+    if (result != ResultOk) {
+        return;
+    }
+
+    const BatchMessageId& msgId = static_cast<const BatchMessageId&>(msg.getMessageId());
+
+    // Only acknowledge on the first message in the batch
+    if (msgId.batchIndex_ <= 0) {
```

Comment:
Shouldn't we acknowledge on the last message of the batch since you clear the Receiver Queue on reconnect

Reply:
It doesn't really makes a lot of difference (and it's easy to ack on the first rather than the last), because for the reader the acknowledge it's just an internal trick that enables us to track the "backlog" in the stats. Since the cursor is non-durable and the reader always specify where to start reading, there is no other meaning for the acks.